lib/posix-process/clone.c

```diff
@@ -0,0 +1,569 @@
+	if (flags & CLONE_PIDFD)		uk_pr_debug(" PIDFD");
+	if (flags & CLONE_PTRACE)		uk_pr_debug(" PTRACE");
+	if (flags & CLONE_VFORK)		uk_pr_debug(" VFORK");
+	if (flags & CLONE_PARENT)		uk_pr_debug(" PARENT");
+	if (flags & CLONE_THREAD)		uk_pr_debug(" THREAD");
+	if (flags & CLONE_NEWNS)		uk_pr_debug(" NEWNS");
+	if (flags & CLONE_SYSVSEM)		uk_pr_debug(" SYSVSEM");
+	if (flags & CLONE_SETTLS)		uk_pr_debug(" SETTLS");
+	if (flags & CLONE_PARENT_SETTID)	uk_pr_debug(" PARENT_SETTID");
+	if (flags & CLONE_CHILD_CLEARTID)	uk_pr_debug(" CHILD_CLEARTID");
+	if (flags & CLONE_DETACHED)		uk_pr_debug(" DETACHED");
+	if (flags & CLONE_UNTRACED)		uk_pr_debug(" UNTRACED");
+	if (flags & CLONE_CHILD_SETTID)		uk_pr_debug(" CHILD_SETTID");
+	if (flags & CLONE_NEWCGROUP)		uk_pr_debug(" NEWCGROUP");
+	if (flags & CLONE_NEWUTS)		uk_pr_debug(" NEWUTS");
+	if (flags & CLONE_NEWIPC)		uk_pr_debug(" NEWIPC");
+	if (flags & CLONE_NEWUSER)		uk_pr_debug(" NEWUSER");
+	if (flags & CLONE_NEWPID)		uk_pr_debug(" NEWPID");
+	if (flags & CLONE_NEWNET)		uk_pr_debug(" NEWNET");
+	if (flags & CLONE_IO)			uk_pr_debug(" IO");
+	uk_pr_debug(" ]\n");
+	if (flags & CLONE_PIDFD)
+		uk_pr_debug(" pidfd: %d\n", (int) cl_args->pidfd);
+	if (flags & CLONE_PARENT_SETTID)
+		uk_pr_debug(" parent_tid: %p\n", (void *) cl_args->parent_tid);
+	if (flags & (CLONE_CHILD_CLEARTID | CLONE_CHILD_SETTID))
+		uk_pr_debug(" child_tid: %p\n", (void *) cl_args->child_tid);
+	uk_pr_debug(" stack: %p\n", (void *) cl_args->stack);
+	uk_pr_debug(" tls: %p\n", (void *) cl_args->tls);
+	uk_pr_debug(" <return>: %p\n", (void *) return_addr);
+	uk_pr_debug(")\n");
+#endif /* UK_DEBUG */
+
+	if ((flags & CLONE_SETTLS)
+#if CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS
+	    && (uk_syscall_ultlsp() == 0x0)
+#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
+	) {
+		/* The caller already created a TLS for the child (for instance
+		 * by a pthread API wrapper). We expect that this TLS is a
+		 * Unikraft TLS.
+		 */
+		uk_pr_debug("Using passed TLS pointer %p as an Unikraft TLS\n",
+			    (void *) cl_args->tls);
+		child = uk_thread_create_container2(s->a,
+						    (__uptr) cl_args->stack,
+						    (__uptr) cl_args->tls,
+						    true, /* TLS is an UKTLS */
+						    false, /* We want ECTX */
+						    (t->name) ? strdup(t->name)
+							      : NULL,
+						    NULL,
+						    _clone_child_gc);
+	} else {
+		/* If no TLS was given or the parent calls us already from
+		 * a context with an userland TLS activated (kernel land vs.
+		 * user land), we allocate an Unikraft TLS because Unikraft
+		 * places TLS variables and uses them effectively as TCB.
+		 */
+#if CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS
+		if (uk_syscall_ultlsp() != 0x0) {
+			uk_pr_debug("Allocating an Unikraft TLS for the new child, parent called from context with custom TLS\n");
+		} else
+#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
+		{
+			uk_pr_debug("Allocating an Unikraft TLS for the new child, no TLS given by parent\n");
+		}
```

Comment:
```suggestion
#if CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS
		if (uk_syscall_ultlsp() != 0x0) {
			uk_pr_debug("Allocating an Unikraft TLS for the new child, parent called from context with custom TLS\n");
		} else {
			uk_pr_debug("Allocating an Unikraft TLS for the new child, no TLS given by parent\n");
		}
#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
```

Reply:
Is this okay, I don't get the placement there

Reply:
My intention here is to compile-in the `if-else` with this configuration option set. If the option is not set, the message of the else block should stay (the other case is not supported without that option).